### PR TITLE
Use sort_by_cached_key when the key function is not trivial/free

### PR DESCRIPTION
**Pull request by ljedrz**
I'm not 100% sure about `def_path_hash` (everything it does is inlined) but it seems like a good idea at least for the rest, as they are cloning.

### CODE REVIEW

File: src/librustc_mir/monomorphize/partitioning.rs [hunk 1]
```diff
@@ -585,7 +585,7 @@ fn merge_codegen_units<'tcx>(tcx: TyCtxt<'_, 'tcx, 'tcx>,
     // smallest into each other) we're sure to start off with a deterministic
     // order (sorted by name). This'll mean that if two cgus have the same size
     // the stable sort below will keep everything nice and deterministic.
-    codegen_units.sort_by_key(|cgu| cgu.name().clone());
+    codegen_units.sort_by_cached_key(|cgu| cgu.name().clone());
```
Comment (sinkuu) [inline src/librustc_mir/monomorphize/partitioning.rs]: Can't it be `sort_by(|a, b| a.name().cmp(b.name()))` to avoid heap allocations completely?

**Comment by nikomatsakis** [comment]
@rust-timer build a22afbf1e892dcdfc1cb62867f6b02fe8e50a700

File: src/librustc/middle/resolve_lifetime.rs [hunk 2]
```diff
@@ -1573,7 +1573,7 @@ impl<'a, 'tcx> LifetimeContext<'a, 'tcx> {
             .collect();
 
         // ensure that we issue lints in a repeatable order
-        def_ids.sort_by_key(|&def_id| self.tcx.def_path_hash(def_id));
+        def_ids.sort_by_cached_key(|&def_id| self.tcx.def_path_hash(def_id));
```
Comment (nikomatsakis) [inline src/librustc/middle/resolve_lifetime.rs]: As you noted, I think this is already cached -- it's something we do a lot in incremental. You agree, @michaelwoerister ?

File: src/librustc/traits/object_safety.rs [hunk 3]
```diff
@@ -408,7 +408,7 @@ impl<'a, 'tcx> TyCtxt<'a, 'tcx, 'tcx> {
             .collect::<Vec<_>>();
 
         // existential predicates need to be in a specific order
-        associated_types.sort_by_key(|item| self.def_path_hash(item.def_id));
+        associated_types.sort_by_cached_key(|item| self.def_path_hash(item.def_id));
```
Comment (nikomatsakis) [inline src/librustc/traits/object_safety.rs]: Also here, of course

File: src/librustc_mir/borrow_check/mod.rs [hunk 4]
```diff
@@ -341,7 +341,7 @@ fn do_mir_borrowck<'a, 'gcx, 'tcx>(
     }
 
     if !mbcx.errors_buffer.is_empty() {
-        mbcx.errors_buffer.sort_by_key(|diag| diag.span.primary_span());
+        mbcx.errors_buffer.sort_by_cached_key(|diag| diag.span.primary_span());
```
Comment (nikomatsakis) [inline src/librustc_mir/borrow_check/mod.rs]: This is the 'error reporting' path, so it really doesn't matter... also `primary_span` is a very simple operation... doesn't strike me as worthwhile.

File: src/librustc_mir/monomorphize/partitioning.rs [hunk 5]
```diff
@@ -985,7 +985,7 @@ fn collect_and_partition_mono_items<'a, 'tcx>(
                 output.push_str(" @@");
                 let mut empty = Vec::new();
                 let cgus = item_to_cgus.get_mut(i).unwrap_or(&mut empty);
-                cgus.as_mut_slice().sort_by_key(|&(ref name, _)| name.clone());
+                cgus.as_mut_slice().sort_by_cached_key(|&(ref name, _)| name.clone());
```
Comment (nikomatsakis) [inline src/librustc_mir/monomorphize/partitioning.rs]: *This* might make sense. Not sure what type `name` has here... :)

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
Let's see what perf says. I think that the `primary_span` and `def_path_hash` calls are likely not a win. Not sure about the others.

**Comment by ljedrz** [comment]
Looks like a win:
```
inflate-check
	avg: -2.4%	min: -3.3%	max: 0.3%
keccak-check
	avg: -2.1%	min: -3.1%	max: 0.5%
keccak-opt
	avg: -1.6%	min: -2.7%	max: 0.4%
keccak-debug
	avg: -1.5%	min: -2.5%	max: 0.3%
inflate-debug
	avg: -1.6%	min: -2.3%	max: 0.3%
```
I can update the commit taking into account the comments - theoretically it could make the wins even nicer.

**Comment by ljedrz** [comment]
Not much was left after including the remarks ^^. We might want to run another round of perf to make sure we didn't lose the gains reported by the first run.

**Comment by nikomatsakis** [comment]
Yeah, we should try again. I'm a bit surprised to see a 3% swing, I have to admit.

File: src/librustc/middle/resolve_lifetime.rs (hunk 2 above)
Comment (michaelwoerister) [reply src/librustc/middle/resolve_lifetime.rs]: It's cached, yes, but still not entirely trivial. I.e. it has to do the local/upstream crate dispatch and for upstream `DefIds` it has to go through a number of function calls: https://github.com/rust-lang/rust/blob/780658a464603fa755d94b27f72a375bd81d07ea/src/librustc/ty/context.rs#L1325

So, I can imagine this being a win for sorting where each key might be compared against multiple times.

**Comment by michaelwoerister** [comment]
The only change left at this point is in debugging code that is only executed by the test suite -- so I expect no wins there. The original changes (except for the `cgu.name().clone()`) looked fine to me though.

**Comment by michaelwoerister** [comment]
lgtm: @bors r+

**Pull request merged**